Automattic/woocommerce-payments
# Allow payment_intent_succeeded webhook to handle orders without intent_id attached

Fixes 1202-gh-Automattic/woocommerce-payments-server

#### Changes proposed in this Pull Request

There has been a lot of reports in different merchants about a situation where an order was successfully paid in Stripe side, but canceled in WooCommerce (because of `woocommerce_hold_stock_minutes`).

Find more about it and its investigation in the shared links here 1202-gh-Automattic/woocommerce-payments-server

This PR allows the webhook to obtain an order using the attached `order_id` in the metadata of the object, in case the intent_id was not properly attached into the order because of any racing condition.

@dmallory42 has also contributed moving `attach_intent_info_to_order` and `attach_exchange_info_to_order` higher in the execution to try to mitigate the racing condition as we spoke here p1634210077172300-slack-C0208C3BXHP

#### Testing instructions

* Run the tests, I've added a new one that gets an order when one is not found using intent_id.
* To test it manually, go to this line https://github.com/Automattic/woocommerce-payments/blob/38dade9f92d20c715aedc89ffdf744f787884066/includes/admin/class-wc-rest-payments-webhook-controller.php#L264 and change it for 
`$order = $this->wcpay_db->order_from_intent_id( "anything" );` in order to simulate getting an intent_id that we had no record of.
* Make a purchase
* See in the logs `intent_id not found, using order_id to retrieve order`, that indicates the webhook has gone into the new added logic.
* Check the order is successfully marked as `payment_complete` after the order was successfully retrieved using the webhook metadata.

-------------------


**Post merge**
